File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        // Create a map of String keys and list values and assign keys of the sorted characters of the string
        // If a character is equal to that sorted character, we add it to that list
        // Return our list
        
        Map<String, List<String>> map = new HashMap<>();
        for(String str : strs){
            char[] array = str.toCharArray();
            Arrays.sort(array);
            map.put(new String(array), new ArrayList<>());
        }
        
        for(String str: strs){
            char[] array = str.toCharArray();
            Arrays.sort(array);
            String sortedString = new String(array);
            if(map.containsKey(sortedString)){
                map.get(sortedString).add(str);
            }
        }
        
        return new ArrayList<>(map.values());
    }
}

File: 0049-group-anagrams/0049-group-anagrams.py
from collections import defaultdict
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        an_map = defaultdict(list)
        for s in strs:
            an_map[self.getAnagram(s)].append(s)
        
        return an_map.values()
        
    
    def getAnagram(self, s):
        c_map = [0] * 26
        
        for c in s:
            i = ord(c) - ord('a')
            c_map[i] += 1
        
        return tuple(c_map)